### DISCUSSION TITLE
Trying to install and run on Ubuntu 21.10

### DISCUSSION THREAD

**Discussion by gsargent70** [sn4k3/UVtools#342] [Q&A] [answered]
Any help with this error once UVtools installed and trying to run on Ubuntu 21.10?

System.DllNotFoundException: Unable to load shared library 'dl' or one of its dependencies

I have installed all dependencies per instructions.  I have not compiled a new version of libcvextern.  One thing to note is I used packages libdc1394-25 and libopenexr25 as these were the only versions available for 21.10.  All other dependencies installed without issue.

Thanks!

**Comment by sn4k3** [comment]
You need to find out what is missing.
On UVtools directory run a terminal and run the command: `ldd libcvextern.so |grep not`

Post the output here

**Reply by gsargent70** [reply]
I get nothing on the display running that.

greg@gregshop:~/Applications$ cd UVtools
greg@gregshop:~/Applications/UVtools$ ldd libcvextern.so | grep not
greg@gregshop:~/Applications/UVtools$ ldd libcvextern.so |grep not
greg@gregshop:~/Applications/UVtools$ 

When I try to run UVtools here is the error strings

greg@gregshop:~/Applications/UVtools$ ./UVtools.sh
Unhandled exception. System.TypeInitializationException: The type initializer for 'Emgu.CV.CvInvoke' threw an exception.
 ---> System.DllNotFoundException: Unable to load shared library 'dl' or one of its dependencies. In order to help diagnose loading problems, consider setting the LD_DEBUG environment variable: libdl: cannot open shared object file: No such file or directory
   at Emgu.Util.Toolbox.Dlopen(String dllname, Int32 mode)
   at Emgu.Util.Toolbox.LoadLibrary(String dllname)
   at Emgu.CV.CvInvoke.LoadUnmanagedModules(String loadDirectory, String[] unmanagedModules)
   at Emgu.CV.CvInvoke.DefaultLoadUnmanagedModules(String[] modules)
   at Emgu.CV.CvInvoke..cctor()
   --- End of inner exception stack trace ---
   at Emgu.CV.CvInvoke.Init()
   at UVtools.WPF.App.OnFrameworkInitializationCompleted() in D:\Tiago\Dropbox\Programming\C#\UVtools\UVtools.WPF\App.axaml.cs:line 70
   at System.Threading.Tasks.Task.<>c.<ThrowAsync>b__140_0(Object state)
   at Avalonia.Threading.AvaloniaSynchronizationContext.<>c__DisplayClass5_0.<Post>b__0() in /_/src/Avalonia.Base/Threading/AvaloniaSynchronizationContext.cs:line 33
   at Avalonia.Threading.JobRunner.Job.Avalonia.Threading.JobRunner.IJob.Run() in /_/src/Avalonia.Base/Threading/JobRunner.cs:line 166
   at Avalonia.Threading.JobRunner.RunJobs(Nullable`1 priority) in /_/src/Avalonia.Base/Threading/JobRunner.cs:line 37
   at Avalonia.X11.X11PlatformThreading.CheckSignaled() in /_/src/Avalonia.X11/X11PlatformThreading.cs:line 164
   at Avalonia.X11.X11PlatformThreading.RunLoop(CancellationToken cancellationToken) in /_/src/Avalonia.X11/X11PlatformThreading.cs:line 244
   at Avalonia.Threading.Dispatcher.MainLoop(CancellationToken cancellationToken) in /_/src/Avalonia.Base/Threading/Dispatcher.cs:line 61
   at Avalonia.Controls.ApplicationLifetimes.ClassicDesktopStyleApplicationLifetime.Start(String[] args) in /_/src/Avalonia.Controls/ApplicationLifetimes/ClassicDesktopStyleApplicationLifetime.cs:line 126
   at Avalonia.ClassicDesktopStyleApplicationLifetimeExtensions.StartWithClassicDesktopLifetime[T](T builder, String[] args, ShutdownMode shutdownMode) in /_/src/Avalonia.Controls/ApplicationLifetimes/ClassicDesktopStyleApplicationLifetime.cs:line 175
   at UVtools.WPF.Program.Main(String[] args) in D:\Tiago\Dropbox\Programming\C#\UVtools\UVtools.WPF\Program.cs:line 47
./UVtools.sh: line 3: 10773 Aborted                 (core dumped) ./UVtools
./UVtools.sh: line 3: dotnet: command not found

**Reply by gsargent70** [reply]
Just for more info here is the output without piping it to grep

greg@gregshop:~/Applications/UVtools$ ldd libcvextern.so
	linux-vdso.so.1 (0x00007ffd7adb6000)
	libgeotiff.so.5 => /lib/x86_64-linux-gnu/libgeotiff.so.5 (0x00007f77172b7000)
	libusb-1.0.so.0 => /lib/x86_64-linux-gnu/libusb-1.0.so.0 (0x00007f771729a000)
	libX11.so.6 => /lib/x86_64-linux-gnu/libX11.so.6 (0x00007f771715c000)
	libXt.so.6 => /lib/x86_64-linux-gnu/libXt.so.6 (0x00007f77170ef000)
	libOpenGL.so.0 => /lib/x86_64-linux-gnu/libOpenGL.so.0 (0x00007f77170c3000)
	libGLX.so.0 => /lib/x86_64-linux-gnu/libGLX.so.0 (0x00007f771708f000)
	libpthread.so.0 => /lib/x86_64-linux-gnu/libpthread.so.0 (0x00007f7717088000)
	libdl.so.2 => /lib/x86_64-linux-gnu/libdl.so.2 (0x00007f7717083000)
	libstdc++.so.6 => /lib/x86_64-linux-gnu/libstdc++.so.6 (0x00007f7716e6a000)
	libm.so.6 => /lib/x86_64-linux-gnu/libm.so.6 (0x00007f7716d86000)
	libc.so.6 => /lib/x86_64-linux-gnu/libc.so.6 (0x00007f7716b5e000)
	/lib64/ld-linux-x86-64.so.2 (0x00007f771d833000)
	libgcc_s.so.1 => /lib/x86_64-linux-gnu/libgcc_s.so.1 (0x00007f7716b44000)
	libtiff.so.5 => /lib/x86_64-linux-gnu/libtiff.so.5 (0x00007f7716abd000)
	libproj.so.19 => /lib/x86_64-linux-gnu/libproj.so.19 (0x00007f7716749000)
	libudev.so.1 => /lib/x86_64-linux-gnu/libudev.so.1 (0x00007f7716720000)
	libxcb.so.1 => /lib/x86_64-linux-gnu/libxcb.so.1 (0x00007f77166f6000)
	libSM.so.6 => /lib/x86_64-linux-gnu/libSM.so.6 (0x00007f77166eb000)
	libICE.so.6 => /lib/x86_64-linux-gnu/libICE.so.6 (0x00007f77166ce000)
	libGLdispatch.so.0 => /lib/x86_64-linux-gnu/libGLdispatch.so.0 (0x00007f7716614000)
	libwebp.so.6 => /lib/x86_64-linux-gnu/libwebp.so.6 (0x00007f77165ac000)
	libzstd.so.1 => /lib/x86_64-linux-gnu/libzstd.so.1 (0x00007f77164df000)
	liblzma.so.5 => /lib/x86_64-linux-gnu/liblzma.so.5 (0x00007f77164b4000)
	libjbig.so.0 => /lib/x86_64-linux-gnu/libjbig.so.0 (0x00007f77162a6000)
	libjpeg.so.8 => /lib/x86_64-linux-gnu/libjpeg.so.8 (0x00007f7716231000)
	libdeflate.so.0 => /lib/x86_64-linux-gnu/libdeflate.so.0 (0x00007f7716216000)
	libz.so.1 => /lib/x86_64-linux-gnu/libz.so.1 (0x00007f77161fa000)
	libsqlite3.so.0 => /lib/x86_64-linux-gnu/libsqlite3.so.0 (0x00007f77160b0000)
	libcurl-gnutls.so.4 => /lib/x86_64-linux-gnu/libcurl-gnutls.so.4 (0x00007f7716017000)
	libXau.so.6 => /lib/x86_64-linux-gnu/libXau.so.6 (0x00007f771600f000)
	libXdmcp.so.6 => /lib/x86_64-linux-gnu/libXdmcp.so.6 (0x00007f7716007000)
	libuuid.so.1 => /lib/x86_64-linux-gnu/libuuid.so.1 (0x00007f7715ffe000)
	libbsd.so.0 => /lib/x86_64-linux-gnu/libbsd.so.0 (0x00007f7715fe6000)
	libnghttp2.so.14 => /lib/x86_64-linux-gnu/libnghttp2.so.14 (0x00007f7715fb8000)
	libidn2.so.0 => /lib/x86_64-linux-gnu/libidn2.so.0 (0x00007f7715f97000)
	librtmp.so.1 => /lib/x86_64-linux-gnu/librtmp.so.1 (0x00007f7715f76000)
	libssh.so.4 => /lib/x86_64-linux-gnu/libssh.so.4 (0x00007f7715f09000)
	libpsl.so.5 => /lib/x86_64-linux-gnu/libpsl.so.5 (0x00007f7715ef5000)
	libnettle.so.8 => /lib/x86_64-linux-gnu/libnettle.so.8 (0x00007f7715eaf000)
	libgnutls.so.30 => /lib/x86_64-linux-gnu/libgnutls.so.30 (0x00007f7715cca000)
	libgssapi_krb5.so.2 => /lib/x86_64-linux-gnu/libgssapi_krb5.so.2 (0x00007f7715c78000)
	libldap-2.5.so.0 => /lib/x86_64-linux-gnu/libldap-2.5.so.0 (0x00007f7715c17000)
	liblber-2.5.so.0 => /lib/x86_64-linux-gnu/liblber-2.5.so.0 (0x00007f7715c06000)
	libbrotlidec.so.1 => /lib/x86_64-linux-gnu/libbrotlidec.so.1 (0x00007f7715bf8000)
	libmd.so.0 => /lib/x86_64-linux-gnu/libmd.so.0 (0x00007f7715beb000)
	libunistring.so.2 => /lib/x86_64-linux-gnu/libunistring.so.2 (0x00007f7715a69000)
	libhogweed.so.6 => /lib/x86_64-linux-gnu/libhogweed.so.6 (0x00007f7715a1f000)
	libgmp.so.10 => /lib/x86_64-linux-gnu/libgmp.so.10 (0x00007f771599c000)
	libcrypto.so.1.1 => /lib/x86_64-linux-gnu/libcrypto.so.1.1 (0x00007f77156c1000)
	libp11-kit.so.0 => /lib/x86_64-linux-gnu/libp11-kit.so.0 (0x00007f7715587000)
	libtasn1.so.6 => /lib/x86_64-linux-gnu/libtasn1.so.6 (0x00007f7715571000)
	libkrb5.so.3 => /lib/x86_64-linux-gnu/libkrb5.so.3 (0x00007f77154a6000)
	libk5crypto.so.3 => /lib/x86_64-linux-gnu/libk5crypto.so.3 (0x00007f7715479000)
	libcom_err.so.2 => /lib/x86_64-linux-gnu/libcom_err.so.2 (0x00007f7715473000)
	libkrb5support.so.0 => /lib/x86_64-linux-gnu/libkrb5support.so.0 (0x00007f7715465000)
	libsasl2.so.2 => /lib/x86_64-linux-gnu/libsasl2.so.2 (0x00007f771544a000)
	libbrotlicommon.so.1 => /lib/x86_64-linux-gnu/libbrotlicommon.so.1 (0x00007f7715427000)
	libffi.so.8 => /lib/x86_64-linux-gnu/libffi.so.8 (0x00007f7715419000)
	libkeyutils.so.1 => /lib/x86_64-linux-gnu/libkeyutils.so.1 (0x00007f7715412000)
	libresolv.so.2 => /lib/x86_64-linux-gnu/libresolv.so.2 (0x00007f77153fd000)

**Reply by sn4k3** [reply]
I think the problem is with libopenexr25, the libcvextern was compiled linking libopenexr24 and with ubuntu 20.x.
You can try compiling the libcvextern as the instructions and see if that helps: https://github.com/sn4k3/UVtools#compile-libcvexternso

**Reply by gsargent70** [reply]
unfortunately same result after doing the compile.  Also it didn't create a emgucv/libs/x64/libcvextern.so file or even that directory structure.

**Reply by sn4k3** [reply]
If libcvextern.so wasn't output, that mean compilation failed in some point. It must output the file to you replace it on UVtools folder

**Reply by gsargent70** [reply]
I'll keep messing with it but seems that it might be better to just run it in a windows VM in linux instead of having to compile and troubleshoot compiling.

**Reply by sn4k3** [reply]
May be hard, until they have official support for new Ubuntu the linking of older libraries will be a problem.
However compilation must tell the error. You can post the last output here from the terminal

**Reply by gsargent70** [reply]
So was able to compile but I had to follow these instructions first so I had the dotnet SDK and runtimes on my machine

https://docs.microsoft.com/en-us/dotnet/core/install/linux-ubuntu

This did allow me to create a new libcvextern.so and copy it to the UVtools folder but this did change anything.  Still the same error messages as above.  I am a little confused by this error message

UVtools.WPF.App.OnFrameworkInitializationCompleted() in D:\Tiago\Dropbox\Programming\C#\UVtools\UVtools.WPF\App.axaml.cs:line 70

Why is this trying to reference a file that is a hard coded path to I assume the development machine you are using?  It references that path again later on in another error.  And the main error says libdl: cannot open shared object file: No such file or directory as in it can't find it.   Is it trying to look at the hard coded path referenced above?

**Reply by sn4k3** [reply]
libcvextern do not require the dotnet sdk, but emgucv does and the make file compile both opencv and emgucv which checks make compile to fail if you dont have dotnet sdk, mono or msbuild.

I can't find any information regarding opencv + 21.10, but the dependencies are a problem. I can try latter my luck on a VM

**Reply by gsargent70** [reply]
Don't know if this helps but in the Syslog I get this error

Trying to load libopencv_videoio_ffmpeg454_64.so using default path.

I have looked everywhere for that library but it doesn't seem to be on the system anywhere.  I verified that I have loaded the libavcodec-dev package.  I did find a very similar issue for another product here throwing the same type of error but no fix yet.

https://www.reddit.com/r/ispyconnect/comments/qmig16/startup_fails_after_ubuntu_2110_upgrade/

Hope this helps.

**Reply by sn4k3** [reply]
do you have the ffmpeg and ffmpeg-dev installed?

**Reply by gsargent70** [reply]
Also found this in the output of the compile for emgucv and opencv.

-- Checking for modules 'libavcodec;libavformat;libavutil;libswscale'
--   No package 'libavcodec' found
--   No package 'libavformat' found
--   No package 'libavutil' found
--   No package 'libswscale' found
-- FFMPEG is disabled. Required libraries: libavcodec;libavformat;libavutil;libswscale. Missing libraries: libavcodec;libavformat;libavutil;libswscale

I assume this means it won't compile the .so file I listed above but I checked and those libraries are installed in my system so don't know why it can't find them.

**Reply by gsargent70** [reply]
> do you have the ffmpeg and ffmpeg-dev installed?

I do not.  Just installed ffmpeg but I have no ffmpeg-dev package

**Reply by sn4k3** [reply]
Ok so now with ffmpeg  try to recompile and repeat the copy of .so

**Reply by gsargent70** [reply]
nope same results with the same errors so that appears not to be it

**Reply by sn4k3** [reply]
Since this is related to opencv you can try to post your problem here: https://github.com/emgucv/emgucv/discussions

**Comment by tslater2006** [comment]
Those aren't hard-coded paths the application is looking for.  They are
part of debug/ stack information for error reporting only. Pretty common
for those paths to match the machine where they were compiled.

**Comment by gsargent70** [comment, marked as answer]
After many hours I have found the solution.  It appears that if you do not have libdl named exactly as libdl.so it can not find it when booting UVtools.  On Ubuntu 21.10 you must create a symlink from your existing libdl to one named as above.  Here is an example command to do this.

sudo ln -s /lib/x86_64-linux-gnu/libdl.so.2 /lib/x86_64-linux-gnu/libdl.so

Once I did this UVtools booted right up from a fresh install.  I did not need to do any recompile of emgucv.  I hope this helps others with their issues.

**Reply by Daoenti** [reply]
This is less a UVTools thing and more a Linux general thing but look into the ldconfig command on your system, it is responsible for maintaining those symlinks in the library directories and should be ran any time a library is installed. Package installs of libraries do this automatically, it’s really only when you have to compile or install a library outside of packages.

**Reply by gsargent70** [reply]
Very strange then as all the libraries and dependencies were installed through the apt package manager so sounds like it should have triggered ldconfig during installs.
